### PR TITLE
bcftools updates for version1.4

### PR DESCRIPTION
**Pull request by jj-umn**
bcftools updated.  New plugins have not yet been added.

### CODE REVIEW

**Comment by jj-umn** [comment]
This contains the new bcftools mpileup command meant to replace the samtools mpileup.  I gave this tool the option of a single input bam/cram or multiple bam/cram.  The single input is to allow a galaxy  workflow to run per chrom variant calling, e.g.:  hisat2 -> bamtools split (update to allow collection output) => bcftools pileup => bcftools call => bcftools merge  -> snpEff

File: tools/bcftools/bcftools_call.xml [hunk 1]
```diff
@@ -250,6 +266,24 @@ SNP/indel variant calling from VCF/BCF. To be used in conjunction with samtools
 
 The novel-rate option can be set to modify the likelihood of novel mutation for constrained -C trio calling. The trio genotype calling maximizes likelihood of a particular combination of genotypes for father, mother and the child P(F=i,M=j,C=k) = P(unconstrained) * Pn + P(constrained) * (1-Pn). By providing three values, the mutation rate Pn is set explicitly for SNPs, deletions and insertions, respectively. If two values are given, the first is interpreted as the mutation rate of SNPs and the second is used to calculate the mutation rate of indels according to their length as Pn=float*exp(-a-b*len), where a=22.8689, b=0.2994 for insertions and a=21.9313, b=0.2856 for deletions [pubmed:23975140]. If only one value is given, the same mutation rate Pn is used for SNPs and indels. 
 
+The prior-freqs workflow looks like this::
+
+    # Extract AN,AC values from an existing VCF, such 1000Genomes
```
Comment (bgruening) [inline tools/bcftools/bcftools_call.xml]: Is this help so useful for users? tabix is not a Galaxy tool for example.

File: tools/bcftools/bcftools_convert_from_vcf.xml [hunk 2]
```diff
@@ -12,14 +12,53 @@
...
+                        <sanitizer sanitize="False"/>
+                    </param>
+                </when>
+            </conditional>
+        </xml>
+        <token name="@SAMPLE_SEX@">
+#if $convert.sex_info.sex_info_src == 'entry':
+  --sex "$entered_sex_file"
+#elif $convert.sex_info.sex_info_src == 'history':
+  --sex "$convert.sex_info.sex_info_file"
```
Comment (bgruening) [inline tools/bcftools/bcftools_convert_from_vcf.xml]: single-quotes

File: tools/bcftools/bcftools_csq.xml [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+@INCLUDE@
+@EXCLUDE@
+@REGIONS@
+@TARGETS@
+
+@OUTPUT_TYPE@
+
+## Primary Input/Outputs
+@INPUT_FILE@
+> "$output_file"
```
Comment (bgruening) [inline tools/bcftools/bcftools_csq.xml]: single-quotes

File: tools/bcftools/bcftools_csq.xml [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+            <expand macro="macro_fasta_ref" />
+            <param name="gff_annot" type="data" format="gff3" label="GFF3 annotation file" 
+                 help="From Ensembl:  ftp://ftp.ensembl.org/pub/current_gff3/"/>
+        </section>
+        <section name="sec_csq_opts" expanded="true" title="CSQ Options">
+            <param name="ncsq" type="integer" value="16" min="1" max="50" label="maximum number of consequences to consider per site"
+                help="-ncsq 16"/>
+            <param name="local_csq" type="boolean" truevalue="--local-csq" falsevalue="" checked="false" label="local-csq"
+                 help="switch off haplotype-aware calling, run localized predictions considering only one VCF record at a time"/>
+            <param name="phase" type="select" optional="true" label="phase"
```
Comment (bgruening) [inline tools/bcftools/bcftools_csq.xml]: Swap label and help?

File: tools/bcftools/bcftools_mpileup.xml [hunk 5]
```diff
@@ -0,0 +1,449 @@
+<?xml version='1.0' encoding='utf-8'?>
+<tool name="bcftools @EXECUTABLE@" id="bcftools_@EXECUTABLE@" version="@VERSION@.0">
+    <description>Generate VCF or BCF containing genotype likelihoods for one or multiple alignment (BAM or CRAM) files</description>
+    <macros>
+        <token name="@EXECUTABLE@">mpileup</token>
+        <import>macros.xml</import>
+        <xml name="bam_flag_options">
+                        <option value="1">Read is paired</option>
+                        <option value="2">Read is mapped in a proper pair</option>
```
Comment (bgruening) [inline tools/bcftools/bcftools_mpileup.xml]: indentation

File: tools/bcftools/bcftools_mpileup.xml [hunk 6]
```diff
@@ -0,0 +1,449 @@
...
+    #set $idx_ext = 'bai'
+    #if $input.input_bam.ext == 'cram':
+        #set $ext = 'cram'
+        #set $idx_ext = 'crai'
+    #end if
+    #set $input_name = $input_base + '.' + $ext
+    #silent $bam_list.append($input_name)
+    ln -s "${input.input_bam}" ${input_name} &&
+    #if $input.input_bam.ext == 'bam':
+      ln -s "${input.input_bam.metadata.bam_index}" ${input_name}.${idx_ext} &&
```
Comment (bgruening) [inline tools/bcftools/bcftools_mpileup.xml]: single quotes

File: tools/bcftools/bcftools_mpileup.xml [hunk 7]
```diff
@@ -0,0 +1,449 @@
...
+        #set $input_base = $re.sub('\W','_',$input_bam.display_name.replace('.bam','').replace('.cram',''))
+        #set $ext = 'bam'
+        #set $idx_ext = 'bai'
+        #if $input_bam.ext == 'cram':
+            #set $ext = 'cram'
+            #set $idx_ext = 'crai'
+        #end if
+        #set $input_name = $input_base + '.' + $ext
+        #silent $bam_list.append($input_name)
+        ln -s "${input_bam}" ${input_name} &&
```
Comment (bgruening) [inline tools/bcftools/bcftools_mpileup.xml]: single quotes

File: tools/bcftools/bcftools_mpileup.xml [hunk 8]
```diff
@@ -0,0 +1,449 @@
...
+    --non-reference
+#end if
+
+#set $section = $sec_indel
+#if $section.perform_indel_calling.perform_indel_calling_selector == 'do_not_perform_indel_calling':
+  --skip-indels
+#elif $section.perform_indel_calling.perform_indel_calling_selector == 'perform_indel_calling':
+  -o "${section.perform_indel_calling.gap_open_sequencing_error_probability}"
+  -e "${section.perform_indel_calling.gap_extension_sequencing_error_probability}"
+  -h "${section.perform_indel_calling.coefficient_for_modeling_homopolymer_errors}"
```
Comment (bgruening) [inline tools/bcftools/bcftools_mpileup.xml]: and here as well :)

File: tools/bcftools/bcftools_mpileup.xml [hunk 9]
```diff
@@ -0,0 +1,449 @@
...
+&& echo 'read-groups:'
+&& cat ${read_groups_file}
+#end if
+
+]]>
+    </command>
+    <inputs>
+
+        <conditional name="input">
+            <param name="input_number" type="select" label="Alignment Inputs">
```
Comment (bgruening) [inline tools/bcftools/bcftools_mpileup.xml]: Why do you need this extra handling? You could just use the multiple-param and user just insert one.

File: tools/bcftools/bcftools_mpileup.xml.20170330 [hunk 10]
```diff
@@ -0,0 +1,434 @@
+<?xml version='1.0' encoding='utf-8'?>
+<tool name="bcftools @EXECUTABLE@" id="bcftools_@EXECUTABLE@" version="@VERSION@.0">
```
Comment (bgruening) [inline tools/bcftools/bcftools_mpileup.xml.20170330]: This file is not needed, correct?

File: tools/bcftools/bcftools_mpileup.xml.20170331 [hunk 11]
```diff
@@ -0,0 +1,197 @@
+<?xml version='1.0' encoding='utf-8'?>
+<tool name="bcftools @EXECUTABLE@" id="bcftools_@EXECUTABLE@" version="@VERSION@.0">
+    <description>Generate VCF or BCF containing genotype likelihoods for one or multiple alignment (BAM or CRAM) files</description>
+    <macros>
```
Comment (bgruening) [inline tools/bcftools/bcftools_mpileup.xml.20170331]: File can be removed?

File: tools/bcftools/bcftools_mpileup.xml.20170331_2 [hunk 12]
```diff
@@ -0,0 +1,272 @@
+<?xml version='1.0' encoding='utf-8'?>
+<tool name="bcftools @EXECUTABLE@" id="bcftools_@EXECUTABLE@" version="@VERSION@.0">
+    <description>Generate VCF or BCF containing genotype likelihoods for one or multiple alignment (BAM or CRAM) files</description>
+    <macros>
+        <token name="@EXECUTABLE@">mpileup</token>
```
Comment (bgruening) [inline tools/bcftools/bcftools_mpileup.xml.20170331_2]: remove this file as well?

File: tools/bcftools/bcftools_roh.xml [hunk 13]
```diff
@@ -39,6 +40,13 @@ ${section.skip_indels}
 #if $section.rec_rate:
   --rec-rate "${section.rec_rate}"
 #end if
+#if $section.buffer_size:
+  #if $section.buffer_overlap:
+    --buffer-size "${section.buffer_size},${section.buffer_overlap}"
```
Comment (bgruening) [inline tools/bcftools/bcftools_roh.xml]: single quotes

File: tools/bcftools/macros.xml [hunk 14]
```diff
@@ -10,9 +10,9 @@
   </xml>
   <xml name="requirements">
     <requirements>
-      <requirement type="package" version="1.3.1">bcftools</requirement>
+      <requirement type="package" version="1.4">bcftools</requirement>
       <!-- conda dependency -->
```
Comment (bgruening) [inline tools/bcftools/macros.xml]: the comment can be removed and the `tool_dependency.xml` file as well.

File: tools/bcftools/macros.xml [hunk 15]
```diff
@@ -279,14 +279,30 @@ $vcfs_list_file
...
     </conditional>
   </xml>
+  <token name="@PREPARE_REGIONS_FILE@">
+<![CDATA[
+#set $regions_path = None
+#if 'regions' in $section
+  #if $section.regions.regions_src == 'regions_file' and $section.regions.regions_file:
+    #if $section.regions.regions_file.ext.startswith('bed'):
+      #set $regions_path = 'regions_file.bed'
+      ln -s "$section.regions.regions_file" $regions_path &&
```
Comment (bgruening) [inline tools/bcftools/macros.xml]: single quotes

File: tools/bcftools/bcftools_convert_from_vcf.xml (hunk 2 above)
Comment (jj-umn) [inline tools/bcftools/bcftools_convert_from_vcf.xml]: @bgruening   Should all file paths be enclosed in single quotes in the command template?

Comment (bgruening) [inline tools/bcftools/bcftools_convert_from_vcf.xml]: Yes, file-path and text parameters should be using single-quotes.
Thanks JJ!

File: tools/bcftools/bcftools_csq.xml [hunk 16]
```diff
@@ -0,0 +1,141 @@
+<?xml version='1.0' encoding='utf-8'?>
+<tool name="bcftools @EXECUTABLE@" id="bcftools_@EXECUTABLE@" version="@VERSION@.0">
+    <description>Haplotype aware consequence predictor</description>
+    <macros>
+        <token name="@EXECUTABLE@">csq</token>
+        <import>macros.xml</import>
+    </macros>
+    <expand macro="requirements">
+        <requirement type="package" version="1.3.1">samtools</requirement>
```
Comment (nsoranzo) [inline tools/bcftools/bcftools_csq.xml]:
```xml
<expand macro="samtools_requirement"/>
```

**Comment by bgruening** [comment]
This file seems to be missing: `test-data/mpileup.3.cram`

**Comment by bgruening** [comment]
@galaxyproject/iuc any help is appreciated this is a large PR and looks really good. Do you spot anything else?

**Comment by jj-umn** [comment]
@bgruening @nsoranzo @erasche   Should we move the bcftools plugins to a subdirectory with a ln to the macros.xml so that they could be a separate repository suite?  
These bcftools plugins do not yet have tool configs: 
- GTisec            
- GTsubset            
- ad-bias            
- af-dist            
- fill-from-fasta            
- fixref            
- guess-ploidy            
- isecGT            
- trio-switch-rate  
I don't know how useful any of them might be, or how difficult it may be to construct tests for them.

**Comment by bgruening** [comment]
@jj-umn not sure it's worth this effort. Admins that are not interested in the plugins can choose the individual tools, isn't it? Or do you think there are many people that want alll bcftools but no plugins?

Are you ok, if we merge this PR or do you want to include the plugins before?

**Comment by jj-umn** [comment]
Do the merge.   I 'll add plugins later as I have time or see a need for them.

**Comment by bgruening** [comment]
Thanks a lot JJ!

**Pull request merged**